x-pack/solutions/security/plugins/security_solution/public/timelines/containers/index.tsx

```diff
@@ -536,19 +528,32 @@ export const useTimelineEvents = ({
      * the combined list of events can be supplied to DataGrid.
      *
      * */
+
+    if (dataLoadingState !== DataLoadingState.loaded) return;
+
+    const { activePage, querySize } = timelineResponse.pageInfo;
+
     setEventsPerPage((prev) => {
-      const result = [...prev];
-      result[timelineResponse.pageInfo.activePage] = timelineResponse.events;
+      let result = [...prev];
+      if (querySize === limit) {
+        result[activePage] = timelineResponse.events;
+      } else {
+        result = [timelineResponse.events];
+      }
       return result;
     });
-  }, [timelineResponse.events, timelineResponse.pageInfo.activePage]);
+  }, [timelineResponse.events, timelineResponse.pageInfo, dataLoadingState, limit]);
 
   useEffect(() => {
     if (!timelineSearchHandler) return;
     timelineSearchHandler();
   }, [timelineSearchHandler]);
 
-  const combinedEvents = useMemo(() => eventsPerPage.flat(), [eventsPerPage]);
+  const combinedEvents = useMemo(
+    // exclude undefined values / empty slots
+    () => eventsPerPage.filter((i) => !!i).flat(),
```

Comment:
nit: `.filter(Boolean)`

Reply:
Thanks for feedback. Done here: https://github.com/elastic/kibana/pull/204034/commits/a599ac6bb2ed0d0b53a2e3e26e577498ae4d3bc0